crawl4urls.py

```diff
@@ -0,0 +1,120 @@
+    paying attention to space out successive hits against same domain."""
+    try:
+        while time() < domain_hit_time.setdefault(url, 0) + MIN_HIT_DELAY:
+            sleep(MIN_HIT_DELAY*0.5)
+        resp = requests.get(url, headers=headers)
+        domain_hit_time[url] = time()
+    except: # Improve: enumerate specific possible exceptions
+        if VERB: print 'Problem downloading',url
+        return None 
+    
+    if resp.status_code != 200: 
+        if VERB: print 'GETing %s generated HTTP error %d' % (url, resp.status_code)
+        return None
+    else: 
+        return StringIO(resp.text)
+
+
+def fetch_html_elem_tree(url):
+    """Wraps *safe* lxml element tree extraction, returning None if not possible"""
+    page = fetch_page(url)
+    if page is None: return
+    
+    try: 
+        tree = etree.parse(page, parser=etree.HTMLParser(encoding='unicode'))
+    except ValueError:
+        return
+    if tree.getroot() is None: return 
+    return tree
+
+
+
+
+class crawl4urls(object):
+    """Basic "breadth first" crawler. 
+    (Implemented as a one-off object 
+    with use syntax as a function for simplicity)"""
+
+    def __init__(self, start_url):
+        init_q = Queue()
+        init_q.put(start_url)
+        self.level_qs = {0: init_q}
+        self.global_todo = set()
+        self.linkedin_urls_seen = set()
+
+        # Start processing link tree level-by-level 
+        # (i.e. "breadth-first")
+        while self.level_qs:
+            self.proc_1_level()
+
+
+    def proc_link(self, url, parent_url):
+        """Prints a LinkedIn profile URL, if found and decides whether
+        to add the link to a list of pages into which to descend 
+        as part of the crawl"""
+        # Unique new LinkedIn url found:
+        if 'linkedin.com/in/' in url and url not in self.linkedin_urls_seen:
+            print '\rLinked in profile found:', url
+            self.linkedin_urls_seen.add(url)            
+        else: # other link...
+            if url[0]=='/': # qualify ref. on local domain
+                scheme, netloc = urlparse(parent_url)[:2]
+                url = '%s://%s%s' % (scheme, netloc, url)
+
+        # If not already on the list of URLs to visit,
+        # add to the queue of the next level of pages to process
+        # (diving in immediately would make this a "depth-first" crawler)
+        if url not in self.global_todo: 
+            self.level_qs.setdefault(self.cur_level+1, Queue()).put(url)
+            self.global_todo.add(url)
+
+
+    def proc_page(self, url):
+        """Process one page - fetching its HTML element tree 
+        and looping through the found <a> tags"""
+        print ('\x1b[2K\rfound: %d  lev: %d  Parsing' \
+                   % (len(self.linkedin_urls_seen), self.cur_level)), url[:100],
+        sys.stdout.flush()
+        tree = fetch_html_elem_tree(url)
+        if tree is None: return
+
+        for link_elem in tree.xpath('//a'):
```

Comment:
The code assumes that a <noscript> block is given as a fall-back for JS-rendered sites, as was the case with the duckduckgo site we started with. (In fact, an earlier version of this code looked for this block first before searching for all <a> tags, but this proved to be redundant.)

Perhaps tweaking the browser agent may yield a more efficient crawl: perhaps some servers use the info to decide on the version of the site to serve.

Honestly, I would need to study Angular and React some more to see if these particular frameworks allow for some trick. Otherwise, I would say we would have to run a JS engine and render the page to crawl it fully.